File: textbook/python/01-dataflow/03_list_filters.py
# Copyright 2026 Terrene Foundation
# SPDX-License-Identifier: Apache-2.0
"""
# ════════════════════════════════════════════════════════════════════════
# Kailash SDK Textbook — DataFlow / List with Filters
# ════════════════════════════════════════════════════════════════════════
# OBJECTIVE: Query records using filters, ordering, limit, and offset
# LEVEL: Basic
# PARITY: Full — Rust uses FilterCondition builder; Python uses dict filters
# VALIDATES: express.list() with filters, order_by, limit, offset
#
# Run: uv run python textbook/python/01-dataflow/03_list_filters.py
# ════════════════════════════════════════════════════════════════════════
"""
from __future__ import annotations

import asyncio
import os
import tempfile

from dataflow import DataFlow

# ── Setup ───────────────────────────────────────────────────────────

db_path = os.path.join(tempfile.gettempdir(), "textbook_dataflow_03.db")
db = DataFlow(database_url=f"sqlite:///{db_path}")


@db.model
class Employee:
    name: str
    department: str
    salary: float
    active: bool = True


async def main():
    await db.initialize()

    # Seed data
    employees = [
        {"name": "Alice", "department": "Engineering", "salary": 95000.0},
        {"name": "Bob", "department": "Engineering", "salary": 85000.0},
        {"name": "Carol", "department": "Marketing", "salary": 78000.0},
        {"name": "Dave", "department": "Marketing", "salary": 72000.0},
        {"name": "Eve", "department": "Engineering", "salary": 110000.0},
        {"name": "Frank", "department": "Sales", "salary": 68000.0, "active": False},
    ]
    for emp in employees:
        await db.express.create("Employee", emp)

    # ── 1. List all ─────────────────────────────────────────────────

    all_emps = await db.express.list("Employee")
    assert len(all_emps) == 6, "Six employees created"

    # ── 2. Filter by exact value ────────────────────────────────────
    # Pass filters as a dict: {field: value}

    engineers = await db.express.list("Employee", filters={"department": "Engineering"})
    assert len(engineers) == 3, "Three engineers"

    active_only = await db.express.list("Employee", filters={"active": True})
    assert len(active_only) == 5, "Five active employees"

    # ── 3. Order by field ───────────────────────────────────────────

    by_salary = await db.express.list("Employee", order_by="salary")
    salaries = [e["salary"] for e in by_salary]
    assert salaries == sorted(salaries), "Ordered by salary ascending"

    # ── 4. Limit and offset (pagination) ────────────────────────────

    page1 = await db.express.list("Employee", order_by="name", limit=2, offset=0)
    assert len(page1) == 2, "Page 1: 2 records"

    page2 = await db.express.list("Employee", order_by="name", limit=2, offset=2)
    assert len(page2) == 2, "Page 2: 2 records"

    # No overlap between pages
    page1_names = {e["name"] for e in page1}
    page2_names = {e["name"] for e in page2}
    assert len(page1_names & page2_names) == 0, "Pages don't overlap"

    # ── 5. Count with filters ───────────────────────────────────────

    eng_count = await db.express.count(
        "Employee", filters={"department": "Engineering"}
    )
    assert eng_count == 3

    total = await db.express.count("Employee")
    assert total == 6

    # ── 6. Combined: filter + order + limit ─────────────────────────

    top_engineers = await db.express.list(
        "Employee",
        filters={"department": "Engineering"},
        order_by="salary",
        limit=2,
    )
    assert len(top_engineers) == 2
    print(f"Top 2 engineers by salary: {[e['name'] for e in top_engineers]}")

    # ── Cleanup ─────────────────────────────────────────────────────
    await db.stop()


asyncio.run(main())

try:
    os.unlink(db_path)
except OSError:
    pass

print("PASS: 01-dataflow/03_list_filters")

File: textbook/rust/01-dataflow/src/bin/03_list_filters.rs
// Copyright 2026 Terrene Foundation
// SPDX-License-Identifier: Apache-2.0
//! Kailash SDK Textbook -- DataFlow / List Filters
//!
//! OBJECTIVE: Use FilterCondition and FilterOp for querying records with conditions.
//! LEVEL: Basic
//! PARITY: Full -- Python uses dict-based filters; Rust uses FilterCondition::new()
//!         with FilterOp enum for type-safe query construction.
//! VALIDATES: FilterCondition, FilterOp, ListOptions with filters
//!
//! Run: cargo run -p tutorial-dataflow --bin 03_list_filters

use kailash_dataflow::prelude::*;
use kailash_value::{Value, ValueMap};
use std::sync::Arc;

#[tokio::main]
async fn main() {
    // ── Setup: Create and Populate ──
    let mut df = DataFlow::new("sqlite::memory:")
        .await
        .expect("connect");

    let model = ModelDefinition::new("Employee", "employees")
        .field("id", FieldType::Integer, |f| f.primary_key())
        .field("name", FieldType::Text, |f| f.required())
        .field("department", FieldType::Text, |f| f.required())
        .field("salary", FieldType::Integer, |f| f.required())
        .field("active", FieldType::Boolean, |f| f.default_value("true"))
        .auto_timestamps();

    df.register_model(model).expect("register");

    let express = df.express();

    // Insert sample employees.
    let employees = vec![
        ("Alice", "Engineering", 90_000, true),
        ("Bob", "Engineering", 85_000, true),
        ("Charlie", "Marketing", 75_000, true),
        ("Diana", "Marketing", 80_000, false),
        ("Eve", "Sales", 70_000, true),
    ];

    for (name, dept, salary, active) in &employees {
        let mut data = ValueMap::new();
        data.insert(Arc::from("name"), Value::from(*name));
        data.insert(Arc::from("department"), Value::from(*dept));
        data.insert(Arc::from("salary"), Value::Integer(*salary));
        data.insert(Arc::from("active"), Value::Bool(*active));
        express.create("Employee", data).await.expect("insert");
    }

    // ── FilterOp::Eq -- Equality ──
    // Filter by exact match on a field value.

    let eng_filter = FilterCondition::new("department", FilterOp::Eq, Value::from("Engineering"));

    let eng = express
        .list(
            "Employee",
            ListOptions {
                filters: vec![eng_filter],
                ..Default::default()
            },
        )
        .await
        .expect("list engineering");
    assert_eq!(eng.len(), 2);

    // ── FilterOp::Gt -- Greater Than ──
    // Salary > 80,000

    let high_salary = FilterCondition::new("salary", FilterOp::Gt, Value::Integer(80_000));

    let rich = express
        .list(
            "Employee",
            ListOptions {
                filters: vec![high_salary],
                ..Default::default()
            },
        )
        .await
        .expect("list high salary");
    assert_eq!(rich.len(), 2); // Alice (90k), Bob (85k)

    // ── FilterOp::Lt -- Less Than ──

    let low_salary = FilterCondition::new("salary", FilterOp::Lt, Value::Integer(80_000));

    let budget = express
        .list(
            "Employee",
            ListOptions {
                filters: vec![low_salary],
                ..Default::default()
            },
        )
        .await
        .expect("list low salary");
    assert_eq!(budget.len(), 2); // Charlie (75k), Eve (70k)

    // ── Combined Filters (AND) ──
    // Multiple filters are combined with AND.

    let active_eng = express
        .list(
            "Employee",
            ListOptions {
                filters: vec![
                    FilterCondition::new("department", FilterOp::Eq, Value::from("Marketing")),
                    FilterCondition::new("active", FilterOp::Eq, Value::Bool(true)),
                ],
                ..Default::default()
            },
        )
        .await
        .expect("list active marketing");
    assert_eq!(active_eng.len(), 1); // Only Charlie (Diana is inactive)

    // ── FilterOp::Gte / FilterOp::Lte -- Range ──
    // Salary >= 75,000 AND salary <= 85,000

    let mid_range = express
        .list(
            "Employee",
            ListOptions {
                filters: vec![
                    FilterCondition::new("salary", FilterOp::Gte, Value::Integer(75_000)),
                    FilterCondition::new("salary", FilterOp::Lte, Value::Integer(85_000)),
                ],
                ..Default::default()
            },
        )
        .await
        .expect("list mid range");
    assert_eq!(mid_range.len(), 3); // Bob (85k), Charlie (75k), Diana (80k)

    // ── FilterOp::Ne -- Not Equal ──

    let not_sales = FilterCondition::new("department", FilterOp::Ne, Value::from("Sales"));
    let non_sales = express
        .list(
            "Employee",
            ListOptions {
                filters: vec![not_sales],
                ..Default::default()
            },
        )
        .await
        .expect("list not sales");
    assert_eq!(non_sales.len(), 4);

    // ── Pagination: limit + offset ──

    let page1 = express
        .list(
            "Employee",
            ListOptions {
                limit: Some(2),
                offset: Some(0),
                ..Default::default()
            },
        )
        .await
        .expect("page 1");
    assert_eq!(page1.len(), 2);

    let page2 = express
        .list(
            "Employee",
            ListOptions {
                limit: Some(2),
                offset: Some(2),
                ..Default::default()
            },
        )
        .await
        .expect("page 2");
    assert_eq!(page2.len(), 2);

    println!("PASS: 01-dataflow/03_list_filters");
}
